### ISSUE TITLE
Servo book on environment setup potentially missing steps for windows 10

### ISSUE THREAD

**Issue by arthmis** [servo/servo#34472]
**Describe the bug:**
This is a bug with the build step documentation for servo on windows 10. 

**To Reproduce:**
I followed the steps to setup the environment [here](https://book.servo.org/hacking/setting-up-your-environment.html#tools-for-windows) and [here](https://github.com/servo/servo?tab=readme-ov-file#windows). However, after executing `.\mach bootstrap` I ran into an error where `crown` the servo linter gave an error of 

```
  Installing crown v0.0.1 (C:\Users\MyName\Projects\servo\support\crown)
   Compiling crown v0.0.1 (C:\Users\MyName\Projects\servo\support\crown)
error: linker `lld-link.exe` not found
  |
  = note: program not found

note: the msvc targets depend on the msvc linker but `link.exe` was not found

note: please ensure that Visual Studio 2017 or later, or Build Tools for Visual Studio were installed with the Visual C++ option.

note: VS Code is a different product, and is not sufficient.
```

I also tried `.\mach build -d` and ran into compilation errors with ring v0.17.8. 
```
warning: ring@0.17.8: Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `clang-cl.exe` installed? (see https://docs.rs/cc/latest/cc/#compile-time-requirements for help)

...some other debug info

  error occurred: Failed to find tool. Is `clang-cl.exe` installed? (see https://docs.rs/cc/latest/cc/#compile-time-requirements for help)
```
So I looked at the [ring docs on how to build it](https://github.com/briansmith/ring/blob/HEAD/BUILDING.md#supported-toolchains-and-cross-compiling) and then did these steps
- installed the Build Tools for Visual Studio 2022
- made sure to install Desktop development with C++ in the installer
- made sure the other servo requirements were installed as well
- then added this `C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\Llvm\x64\bin` to my path so the build tools binaries are available

Adding the build tools binaries to my path also solved the issue with building `crown` in the bootstrap command

After this I was able to build ring during `.\mach build -d` and hit another roadblock at `mozangle v0.5.2`. 
```
thread 'main' panicked at C:\Users\MyName\.cargo\registry\src\index.crates.io-6f17d22bba15001f\bindgen-0.69.5\lib.rs:622:31:
  Unable to find libclang: "couldn't find any valid shared libraries matching: ['clang.dll', 'libclang.dll'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"
```
The error I got suggested that I change the environment variable `LIBCLANG_PATH` to the location of either `libclang.dll` or `clang.dll`. So I added the path to `libclang.dll` found in `C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\Llvm\x64\bin` and was able to compile successfully.

**Platform:**
Microsoft Windows 10 Version 10.0.19045 Build 19045
chocolatey - 2.4.0
python - 3.12
rustup - 1.27.1
Visual Studio Build Tools 2022

So I learned that you don't need to install Visual Studio Community Edition. The [build tools](https://visualstudio.microsoft.com/visual-cpp-build-tools/) are fine. None of the environmental setup steps actually adds the necessary binaries to the path, at least on my install. Does the book need to be updated or is this just an issue on my end?

**Comment by mrobinson** [comment]
It sounds like LLVM did not install correctly during bootstrap. I think the issue isn't missing steps, but we should figure out why LLVM didn't install correctly via chocolatey for you.

**Comment by arthmis** [comment]
Yeah LLVM issue was an issue, but when trying to compile the ring crate it would have failed even if LLVM installed correctly. I use rust normally, on Windows, but I wouldn't have had the `Desktop development with C++` installed using the visual studio installer. Rust doesn't require that for installation. Without that `.\mach build` would fail. Unless when installing `LLVM` through chocolatey those steps would have been done?

I uninstalled everything and tried to follow the servo docs, so I uninstalled rust, rustup, cargo, visual studio tools and chocolatey.

- I did the quick installation with rustup which installed
    - MSVC v143 - VS 2022 C++ x64/x86 build tools (Latest)
    - Windows 11 SDK (10.0.22000.0) - I'm on windows 10
- Then I installed the other requirements in the visual studio installer that servo needs
- I changed directory to servo and ran cargo clean to start fresh which did this step
```
info: syncing channel updates for '1.80.1-x86_64-pc-windows-msvc'
info: latest update on 2024-08-08, rust version 1.80.1 (3f5fd8dd4 2024-08-06)
info: downloading component 'cargo'
info: downloading component 'clippy'
info: downloading component 'llvm-tools'
info: downloading component 'rust-docs'
info: downloading component 'rust-std'
info: downloading component 'rustc'
info: downloading component 'rustc-dev'
info: downloading component 'rustfmt'
```
then I executed `.\mach bootstrap`
I ran into the LLVM issue with chocolatey and finally figured it out. I had an installation of LLVM 10.0.0 years ago and I guess I deleted the folder without running the uninstallation and chocolatey was trying to run the uninstallation exe of the old LLVM which caused the issues. So I reinstalled version `10.0.0` and then uninstalled it the right way. After that chocolatey was able to install LLVM and continue the bootstrap process. So after addressing the LLVM issue bootstrap was successful.

Now I ran `.\mach build -d` and ran into a compilation error 
```
error: linker `lld-link.exe` not found
```
Which I guess is because if there is a `lld linker` it isn't in my path and none of the steps add it. This failure happened while compiling `shlex v1.3.0`

**Comment by arthmis** [comment]
Ok I tried reinstalling everything from scratch and I figured out the issue with my LLVM installation. I had an old version `10.0.0` installed previously which I didn't uninstall correctly. Basically the uninstallation executable was still there but everything else was deleted. Chocolatey, when install LLVM, was trying to run the uninstallation but it kept failing. 
- So I reinstalled `10.0.0` and uninstalled it correctly this time. 
- After that chocolatey could install LLVM correctly.

I was able to run `.\mach bootstrap` which finished successfully and then I ran `.\mach build -d` which did have an issue where it couldn't see the lld linker. I refreshed the environment variables for the shell and it worked correctly. So the compilation was successful.

The odd is part is that, while previously choco couldn't uninstall the old version of LLVM, it still installed the new one successfully, so why was the environment variable set so the build step could see it? Or was there some kind of conflict. I can't go back to see what my old environment variables were like.

**Comment by parasyte** [comment]
I just hit this same issue (Windows 11).

The reason (I think) is that I'm using bash from git-for-windows, and it runs `choco` in a new PowerShell window. I have to manually close the window when it's complete. That's when I hit this error with `lld-link`.

The solution was restarting my shell to get all of the environment updates after `choco` finished. After that, `./mach bootstrap` just worked.

**Comment by jschwe** [comment]
We have switchted to using `winget` now, so going to close this issue (it's entirely possible that a similar issue might exist with winget, but in that case a fresh issue might be helpful)
